Automattic/jetpack
# Settings: Use Constants to check value of REST_API_REQUEST inside the is_syncing function

Fixes https://github.com/Automattic/wp-calypso/issues/76343

## Proposed changes:

We want to render the `jetpack/contact-form` block via the rest API endpoint but it's disabled since it will check the value of [Settings::is_syncing](https://github.com/Automattic/jetpack/blob/fb116496543a656cde086d4bdbd37c3817e783e6/projects/packages/forms/src/contact-form/class-contact-form.php#L243). However, the value is always true if it's in the context of the [REST API request](https://github.com/Automattic/jetpack/blob/fb116496543a656cde086d4bdbd37c3817e783e6/projects/packages/sync/src/class-settings.php#L511).

So, this PR is proposing to use `Constants` class from Jetpack to check the defined value and then we're able to overwrite the value to bypass the check of `Settings::is_syncing` to have a chance to render `jetpack/contact-form` block via the REST API request.

We can also add a new filter, `jetpack_sync_settings_is_syncing`, to filter the value of `Settings::is_syncing`  as below. I'm also fine with this way. WDYT?

```php
public static function is_syncing() {
  $is_syncing = self::$is_syncing || ( defined( 'REST_API_REQUEST' ) && REST_API_REQUEST );
  return (bool) apply_filters( 'jetpack_sync_settings_is_syncing', $is_syncing );
}
```

### Screenshots

| Before | After |
| - | - |
| ![image](https://user-images.githubusercontent.com/13596067/235635701-e223b365-bbee-465d-9d1e-9f46dab0de94.png) | ![image](https://user-images.githubusercontent.com/13596067/235635589-d29f7cfa-bc64-4c0f-a813-37efe7af83d0.png) |
| ![image](https://user-images.githubusercontent.com/13596067/235635848-7f41b15d-c655-4d53-8e0d-0df1bd257286.png) | ![image](https://user-images.githubusercontent.com/13596067/235635796-878e2992-765d-44bf-aa37-c68c938b0cf6.png) |
| ![image](https://user-images.githubusercontent.com/13596067/235636012-99efa23d-6443-4362-863d-c5ff2501d25b.png) | ![image](https://user-images.githubusercontent.com/13596067/235635924-60d75c8b-2ade-4783-92aa-ba4083b61f67.png) |
| ![image](https://user-images.githubusercontent.com/13596067/235636092-0a965ad3-def0-48ff-a7ac-79bdd469e63b.png) | ![image](https://user-images.githubusercontent.com/13596067/235636135-6ebe4426-80b7-4486-9aac-eb2c8e2ce8f6.png) |

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Apply D109531-code to your sandbox
* Apply this PR to your sandbox as below
  * `bin/jetpack-downloader test jetpack feat/rest-api-jetpack-block`
* Go to /setup/site-setup?siteSlug=<your_free_site>
* Continue until you land on the Design Picker
* On the Design Picker screen, scroll down to the bottom and select BCPA CTA
* On the Pattern Assembler screen
  * Select Sections > Add patterns > Contact
  * Ensure the patterns with the contact form are rendered correctly
